libndt.cpp

```diff
@@ -1123,13 +1134,89 @@ bool Client::resolve(const std::string &hostname,
       result = false;
       break;
     }
-    addrs->push_back(address); // we only care about address
+    addrs->push_back(address);  // we only care about address
     EMIT_DEBUG("- " << address);
   }
   this->freeaddrinfo(rp);
   return result;
 }
 
+// Utilities (SSL)
+
+int Client::maybessl_connect(const std::string &hostname, Socket fd,
+                             const sockaddr *sa, SockLen len) noexcept {
+  auto rv = this->connect(fd, sa, len);
+  if (rv == 0 && (impl->settings.proto & protocol::tls) != 0) {
+#ifdef HAVE_OPENSSL
+    SSL *ssl = nullptr;
+    {
+      SSL_CTX *ctx = ::SSL_CTX_new(SSLv23_client_method());
+      if (ctx == nullptr) {
+        EMIT_WARNING("SSL_CTX_new() failed");
+        this->closesocket(fd);
+        return -1;
+      }
+      if (impl->settings.ca_bundle_path.empty()) {
+        EMIT_WARNING("maybessl_connect: ca_bundle_path is empty");
+        ::SSL_CTX_free(ctx);
+        this->closesocket(fd);
+        return -1;
+      }
+      if (!SSL_CTX_load_verify_locations(  //
+              ctx, impl->settings.ca_bundle_path.c_str(), nullptr)) {
+        EMIT_WARNING("SSL_CTX_load_verify_locations() failed");
+        ::SSL_CTX_free(ctx);
+        this->closesocket(fd);
+        return -1;
+      }
+      EMIT_DEBUG("SSL_CTX created");
+      ssl = ::SSL_new(ctx);
+      if (ssl == nullptr) {
+        EMIT_WARNING("SSL_new() failed");
+        ::SSL_CTX_free(ctx);
+        this->closesocket(fd);
+        return -1;
+      }
+      EMIT_DEBUG("SSL created");
+      ::SSL_CTX_free(ctx);
+      impl->fd_to_ssl[fd] = ssl;
+    }
+    {
+      if (!SSL_set_fd(ssl, fd)) {
+        EMIT_WARNING("SSL_set_fd() failed");
+        this->closesocket(fd);
+        return -1;
+      }
+      SSL_set_connect_state(ssl);
+      EMIT_DEBUG("SSL bound to socket");
+    }
+    {
+      // TODO(bassosimone): the following assumes OpenSSL v1.1 or greater. It'd
+      // be better to have a solution also for LibreSSL and BoringSSL.
+      SSL_set_hostflags(ssl, X509_CHECK_FLAG_NO_PARTIAL_WILDCARDS);
+      if (!SSL_set1_host(ssl, hostname.c_str())) {
+        EMIT_WARNING("SSL_set1_host() failed");
+        this->closesocket(fd);
+        return -1;
```

Comment:
Rather than using an OpenSSL 1.1-only pattern, we should use a more broadly compatible pattern, as documented on [the OpenSSL wiki](https://wiki.openssl.org/index.php/Hostname_validation). That should work with old OpenSSL and other SSL flavours.